### PR TITLE
Zoom integration and major architecture upgrades

### PR DESCRIPTION
**Pull request by Nico-J**
Good morning,

This PR introduces a new integration for Zoom Video Meetings. Also I improved some structures and architectures that would become redundant when scaling up the system's complexity.

The following things have been changed:
- Set ENV variables for ZOOM_CLIENT_ID and ZOOM_CLIENT_SECRET
- Improved calendarClient by optimizing the flow of generic "createEvent" (line 334) and generic "updateEvent" (line 353) by using async functions and awaits and newly introduced mail objects to send mails to owner and attendees (see below for more)
- Introduced the abstract base class "EventMail" that now is the very base of every email sent to attendees and event owners when booking, updating oder deleting an event. Previously, there were multiple mail templates, all with their own sending logic and redundant functions. Now, this base class includes all important logic and provides abstract methods when there are things that need to be set individually by each inheriting class. Even the cancel and reschedule link footer is already generated here and just referenced in the sub classes.
- The class EventAttendeeMail extends EventMail and implements all necessary HTML content for the email. In Addition, it provides the getAdditionalBody() method in order to allow inheriting classes to inject more content to the mail body without re-implementing and thus copying the HTML content again. Also, an additionalFooter is implemented in order to append (for example) cancel and reschedule links.
- The class EventAttendeeRescheduledMail extends EventAttendeeMail and is used to tell attendees that their booking has been rescheduled. Currently it shares the getInviteeStart() method of the parent class.
- The class EventOwnerMail extends EventMail and is designed to be sent to the EventOwner when an event has been scheduled. It also includes the iCal string generation and provides it to its subclasses for reusability. The HTML representation of this mail also includes additional bodies and footers.
- Similar to EventAttendeeRescheduledMail, the class EventOwnerRescheduledMail (extending EventOwnerMail) is used to notify an EventOwner when an event has been rescheduled. Everything that would be duplicated is again reused from the parent class.
- In addition, there are the two classes VideoEventAttendeeMail (extending EventAttendeeMail) and VideoEventOwnerMail (extending EventOwnerMail) that provide an additional body containing video call information (provider, meeting id, password, url).
- confirm-booked.ts and new-event.ts, the old, redundantly implemented mail templates, have been deleted, because they are no longer needed.
- A helpers.ts in the /lib/mail folder provides common methods for formatting video call IDs and video integration names. It is used by both VideoEventAttendeeMail and VideoEventOwnerMail. Because they both inherit from different, non-video classes, it cannot be delivered by parental classes.
- Added "zoom_video" to the integrations.
- Added videoClient.ts that, similarly to calendarClient.ts, implements an interface for video integration. For the beginning, it includes Zoom (authentication, availability check, adding events, updating events, deleting events).
- Split up the booking logic into two integration lanes: calendar and video in order to use both integration clients. **I highly recommend to implement the calendarClient.ts and videoClient.ts to a class based hierarchy, implementing a _Strategy Pattern_ to allow an unified handling of the event bookings again. Would be much much cleaner and way more elegant. And highly scalable.**
- Extended the availability check logic to check both integration types (calendar and video). Again, the suggested strategy pattern mentioned above would make this a lot simpler, more scalable and less subject to errors.
- HashUID is now generated in the integration clients, not in the request handler.
- Emails to attendees and owners are now being sent consequently from one point at the calendarClient and videoClient: when creating or updating events. Before this, parts of that were sent in the client and the other part in the booking request handler. **Legacy/backwards compatibility: To still allow users without integrations to send attendee emails, I still generated the HashUID as a fallback in the request handler and send an attendee email, even without any integrations. I recommend, when the strategy pattern is implemented, to implement a MailClientStrategy and quickly get rid of this legacy code.** 
- Added OAuth handling for zoom.
- Changed README.md to contain a manual on how to add the Zoom integration to your Zoom account.
- Several minor bug fixes

### CODE REVIEW

File: pages/integrations/index.tsx [hunk 1]
```diff
@@ -220,6 +220,13 @@ export async function getServerSideProps(context) {
         title: "Office 365 / Outlook.com Calendar",
         imageSrc: "integrations/office-365.png",
         description: "For personal and business accounts",
+    }, {
+        installed: !!(process.env.ZOOM_CLIENT_ID && process.env.ZOOM_CLIENT_SECRET),
+        type: "zoom_video",
+        credential: credentials.find( (integration) => integration.type === "zoom_video" ) || null,
+        title: "Zoom",
+        imageSrc: "integrations/zoom.png",
+        description: "For personal and business accounts",
```
Comment (emrysal) [inline pages/integrations/index.tsx:229]: @PeerRich Do you think this is the best description to use for Zoom?

**Review by emrysal — APPROVED** [review]
As a style, I prefer composition over inheritance, but I agree that this is an improvement over what was before, so it's fine by me (albeit I wouldn't be too happy to apply this everywhere throughout the backend for now). I think email will prove to be fairly stable using this design so the inheritance model may be the right tool for the job here.

File: pages/integrations/index.tsx (hunk 1 above)
Comment (PeerRich) [reply pages/integrations/index.tsx:229]: I would just do "Video Conferencing", at least that's what Zoom says on their website

File: pages/[user]/book.tsx [hunk 2]
```diff
@@ -76,7 +76,7 @@ export default function Book(props) {
             }
         );
 
-        let successUrl = `/success?date=${date}&type=${props.eventType.id}&user=${props.user.username}&reschedule=1`;
+        let successUrl = `/success?date=${date}&type=${props.eventType.id}&user=${props.user.username}&reschedule=${!!rescheduleUid}`;
```
[the remark this answers is not in the document]
Comment (Nico-J) [reply pages/[user]/book.tsx:79]: No, cancellations are handled by /cancel/<uid>, so rescheduleUid is the right term here. :)

File: lib/emails/EventOwnerRescheduledMail.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import dayjs, {Dayjs} from "dayjs";
+import EventOwnerMail from "./EventOwnerMail";
+
+export default class EventOwnerRescheduledMail extends EventOwnerMail {
```
[the remark this answers is not in the document]
Comment (Nico-J) [reply lib/emails/EventOwnerRescheduledMail.ts:4]: When an event is rescheduled from the integration side (e.g. I move an event in my Google Calendar), there is no hook on our side that is called afaik. So our backend would never know and so the attendees wouldn't be notified at all. So I guess, the way to go right now is to allow organizers to reschedule events by those links as well.

Also, @PeerRich asked me to do so (to put the links in the event descriptions as well, which is then being sent to the organizer).

[the remark this answers is not in the document]
Comment (Nico-J) [reply lib/emails/EventOwnerRescheduledMail.ts:4]: And I'm completely with you regarding the naming. Just pushed a commit containing those refactorings.

**Comment by Nico-J** [comment]
Let's first merge the other PRs that touch the calendarClient and the mail templates. All of these would break otherwise when merging this one. I'll then take care of the conflicts in this PR. :)

**Review by PeerRich — APPROVED** [review]

**Comment by PeerRich** [comment]
> Let's first merge the other PRs that touch the calendarClient and the mail templates. All of these would break otherwise when merging this one. I'll then take care of the conflicts in this PR. :)

sounds good. lets do it!

**Comment by baileypumfleet** [comment]
Excellent. All of the other PRs have been merged, so should be all clear to have a look at the conflicts. Let me know if you need anything else!

**Review by emrysal — APPROVED** [review]

**Pull request merged**